overlord/snapstate/snapstate.go

```diff
@@ -583,6 +589,12 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 	addTask(linkSnap)
 	prev = linkSnap
 
+	if snapsup.Type == snap.TypeKernel {
+		setupKernel := st.NewTask("cleanup-old-kernel-snap", fmt.Sprintf(i18n.G("Cleanup kernel driver tree for %q%s"), snapsup.InstanceName(), revisionStr))
+		addTask(setupKernel)
+		prev = setupKernel
+	}
```

Comment:
can we do this this way now that we reboot only once? is the kernel the last thing we do  before reboot? will that stay this way?

Reply:
As far as I understand, tasks for the kernel after "link-snap" should happen after the reboot as the link task would create a restart boundary. If there are other snaps being installed at the same time that require a reboot, those will progress until their restart boundaries, but any task after the boundaries will happen only after the reboot. @Meulengracht is my understanding correct?

Reply:
@alfonsosanchezbeato I'll give this PR a proper review later, but just to answer your question and clear up this. There is a minor problem here due to how edges are used.

In reboot.go (under snapstate), this is mentioned
```
// - BeginEdge (marks the first task of a task-set)
// - EndEdge (marks the last task of a task-set)
// - MaybeRebootEdge (marks the last task that needs to run before the reboot)
// - MaybeRebootWaitEdge (marks the first task that needs to run after the reboot)
```

Currently link-snap is marked MaybeRebootEdge, and auto-connect MaybeRebootWaitEdge. So if you introduce a new task between link-snap and auto-connect, then you *must* modify the edges. Just doing this may not be enough, and I can't give a clear answer just on the top of my head as there is some inherent logic coupling between link-snap and auto-connect as auto-connect does some reboot validation code, so injecting a new task between those two may be an issue. Specifically this is mentioned in the handler code

```

	// The previous task (link-snap) may have triggered a restart,
	// if this is the case we can only proceed once the restart
	// has happened or we may not have all the interfaces of the
	// new core/base snap.
	if err := snapstateFinishRestart(task, snapsup); err != nil {
		return err
	}
```

Reply:
Thanks for explaining - to minimize the changes I will simply move the clean-up of the previous kernel snap so it happens after "auto-connect", as when this task happens exactly is not critical as long as it is done after the restart.